circuitbreaker/circuit_breaker.go

```diff
@@ -54,82 +56,81 @@ type CircuitBreaker struct {
+// This is a blocking function.
+func (cb *CircuitBreaker) Execute(cmd Command) CommandResult {
 	var result CommandResult
+	ctx := cmd.ctx
+	if ctx == nil {
+		ctx = context.Background()
+	}
 
-	for i := 0; i < len(cmd.functors); i += 2 {
-		f1 := cmd.functors[i]
-		var f2 *Functor
-		if i+1 < len(cmd.functors) {
-			f2 = cmd.functors[i+1]
-		}
+	for i := 0; i < len(cmd.functors); i += 1 {
+		f := cmd.functors[i]
 
-		circuitName := fmt.Sprintf("%s_%d", eh.config.CommandName, i)
+		circuitName := f.circuitName
+		if circuitName == "" {
+			circuitName = cb.config.CircuitName
+		}
 		if hystrix.GetCircuitSettings()[circuitName] == nil {
 			hystrix.ConfigureCommand(circuitName, hystrix.CommandConfig{
-				Timeout:                eh.config.Timeout,
-				MaxConcurrentRequests:  eh.config.MaxConcurrentRequests,
-				RequestVolumeThreshold: eh.config.RequestVolumeThreshold,
-				SleepWindow:            eh.config.SleepWindow,
-				ErrorPercentThreshold:  eh.config.ErrorPercentThreshold,
+				Timeout:                cb.config.Timeout,
+				MaxConcurrentRequests:  cb.config.MaxConcurrentRequests,
+				RequestVolumeThreshold: cb.config.RequestVolumeThreshold,
+				SleepWindow:            cb.config.SleepWindow,
+				ErrorPercentThreshold:  cb.config.ErrorPercentThreshold,
 			})
 		}
 
-		// If circuit is the same for all functions, in case of len(cmd.functors) > 2,
-		// main and fallback providers are different next run if first two fail,
-		// which causes health issues for both main and fallback and ErrorPercentThreshold
-		// is reached faster than it should be.
-		errChan := hystrix.Go(circuitName, func() error {
-			res, err := f1.Exec()
-			// Write to resultChan only if success
-			if err == nil {
-				resultChan <- CommandResult{res: res, err: err}
-			}
-			return err
-		}, func(err error) error {
-			// In case of concurrency, we should not execute the fallback
-			if f2 == nil || err == hystrix.ErrMaxConcurrency {
-				return err
-			}
-			res, err := f2.Exec()
+		err := hystrix.DoC(ctx, circuitName, func(ctx context.Context) error {
+			res, err := f.exec()
+			// Write to result only if success
 			if err == nil {
-				resultChan <- CommandResult{res: res, err: err}
+				result = CommandResult{res: res, err: err}
 			}
 			return err
-		})
+		}, nil)
 
-		select {
-		case result = <-resultChan:
-			if result.err == nil {
-				return result
-			}
-		case err := <-errChan:
-			result = CommandResult{err: err}
+		if err == nil {
+			break
+		}
 
-			// In case of max concurrency, we should delay the execution and stop iterating over fallbacks
-			// No error unwrapping here, so use strings.Contains
-			if strings.Contains(err.Error(), hystrix.ErrMaxConcurrency.Error()) {
-				return result
-			}
+		// Accumulate errors
+		if result.err != nil {
+			result.err = fmt.Errorf("%w, %s.error: %w", result.err, circuitName, err)
+		} else {
+			result.err = fmt.Errorf("%s.error: %w", circuitName, err)
 		}
+		result.res = nil
```

Comment:
Don't we need to add this line as well `result.err = nil` otherwise we're referring to the error from the previous functor not the last one?

Reply:
It depends how you implement body in `err := hystrix.DoC(ctx, circuitName, func(ctx context.Context) error {.....`